server/src/main/java/org/candlepin/async/JobArguments.java

```diff
@@ -14,176 +14,185 @@
+     *
+     * @param key
+     *  the key for which to fetch the associated serialized value
+     *
+     * @return
+     *  the serialized value associated with the specified key, or null if the key is not currently
+     *  associated with a value
      */
-    @Override
-    public Collection<Object> values() {
-        return this.map.values();
+    public String getSerializedValue(String key) {
+        return this.data.get(key);
     }
 
     /**
-     * {@inheritDoc}
+     * Fetches the value associated with the specified key, deserialized as the given type. If the
+     * key is not currently associated with any value, this method returns the given default value.
+     *
+     * @param key
+     *  the key for which to fetch the associated value
+     *
+     * @param type
+     *  the type to use to deserialize the the value
+     *
+     * @param defaultValue
+     *  the value to return if the key is not currently associated with a value
+     *
+     * @return
+     *  the deserialized value associated with the specified key, or the given default value if the
+     *  key is not currently associated with a value
      */
-    @Override
-    public String toString() {
-        StringBuilder builder = new StringBuilder(this.getClass().getName());
-
-        builder.append(" {");
-        Iterator<Map.Entry<String, Object>> iterator = this.map.entrySet().iterator();
-        while (iterator.hasNext()) {
-            Map.Entry<String, Object> entry = iterator.next();
-
-            builder.append(entry.getKey());
-            builder.append('=');
-            builder.append(entry.getValue());
-
-            if (iterator.hasNext()) {
-                builder.append(", ");
+    public <T> T getAs(String key, Class<T> type, T defaultValue) {
+        if (this.containsKey(key)) {
+            try {
+                String json = this.data.get(key);
+                return json != null ? mapper.readValue(json, type) : null;
+            }
+            catch (IOException e) {
+                String errmsg = String.format("Unable to deserialize key \"%s\" as type: %s", key, type);
+                throw new ArgumentConversionException(errmsg, e);
             }
         }
-        builder.append('}');
 
-        return builder.toString();
+        return defaultValue;
     }
 
-    // Utility methods
+    /**
+     * Fetches the value associated with the specified key, deserialized as the given type. If the
+     * key is not currently associated with any value, this method returns null.
+     *
+     * @param key
+     *  the key for which to fetch the associated value
+     *
+     * @param type
+     *  the type to use to deserialize the the value
+     *
+     * @return
+     *  the deserialized value associated with the specified key, or null if the key is not
+     *  currently associated with a value
+     */
+    public <T> T getAs(String key, Class<T> type) {
```

Comment:
The class is being passed into the object mapper. That it happens to also be used for type inference is a happy coincidence.

There are shenanigans we can do with reflection to determine the type at runtime, but it's slow and a lot of extra work just to avoid six extra characters.

Reply:
Fair enough.